compiler/qsc/benches/eval.rs

```diff
@@ -13,38 +13,41 @@ const LARGE: &str = include_str!("./large.qs");
 pub fn teleport(c: &mut Criterion) {
     c.bench_function("Teleport evaluation", |b| {
         let sources = SourceMap::new([("Teleportation.qs".into(), TELEPORT.into())], None);
-        let evaluator = stateless::Interpreter::new(true, sources).expect("code should compile");
+        let mut evaluator =
+            stateful::Interpreter::new(true, sources, PackageType::Exe, TargetProfile::Full)
+                .expect("code should compile");
         b.iter(move || {
             let mut out = Vec::new();
             let mut rec = GenericReceiver::new(&mut out);
-            let mut eval_ctx = evaluator.new_eval_context();
-            assert!(eval_ctx.eval_entry(&mut rec).is_ok());
+            assert!(evaluator.eval_entry(&mut rec).is_ok());
```

Comment:
Hmm, there's a behavior difference here (and in the below two benchmarks) because we're not using a fresh evaluation context each time, meaning the iterations aren't as independent as they used to be. Intentional?

Reply:
Yes, that's technically true. In this case we know it's safe to repeatedly run the entry point callable without fully re-initializing the state for a couple of reasons:
- The execution succeeds, so no early termination leaves the environment in any intermediate states.
- The entry point is being invoked directly without any top-level statements, so there are no global variables to contend with and all qubits have been released.
- Released qubits must have been in the |0⟩ state or they would have triggered an error, so there is no leftover global quantum state.
- Insider knowledge of the simulator tells us that when all qubits are released, all accumulated phase is cleared because the internal data structures are reinitialized (see https://github.com/qir-alliance/qir-runner/blob/main/sparsesim/src/lib.rs#L165-L169)

Reply:
You've convinced me it's safe, but is it a valid performance benchmark? Are the iterations sharing any cached state?

Reply:
Fair question. The intent of the benchmark separating out the construction of the interpreter from the evaluation, even in the old API pattern, was to perform compilation outside of the benchmarked loop and focus on measuring just execution time. I don't think we have any shared state between iterations that invalidates that goal; technically the environment object and simulator object are preserved but empty, so the only thing we aren't measuring is the time needed to construct those. That lines up with my intent, at least, of trying as much as possible to measure the perf of the evaluator internals. And the fact that the benchmark numbers stayed the same supports the idea that we are still measuring those internals which are unchanged.

All that said, I think what I outlined above undercuts the use of these benchmarks to measure any change in behavior for interpreter elements that surround the evaluator, because those aren't being measured here. The benchmarks in this file measure just the evaluator, the benchmarks in large.rs and library.rs measure the compiler, but we don't actually have benchmarks that measure the end-to-end of combining the two. Perhaps the wasm folder would be the right place to add those?